src/token_retriever/l1_retriever.rs

```diff
@@ -0,0 +1,142 @@
+use std::fmt;
+
+use http::{header::CONTENT_TYPE, StatusCode, Uri};
+use serde_json::json;
+
+use crate::{
+    authenticator::TokenRetrievalResponse,
+    http_client::{HttpClient, HttpClientError},
+    system_identity::input_data::auth_method::ClientSecret,
+    token::Token,
+    TokenRetriever, TokenRetrieverError,
+};
+
+/// HTTP-based token retriever for L1 authentication method (client ID + client secret)
+pub struct L1TokenRetriever<C: HttpClient> {
+    client_id: String,
+    client_secret: ClientSecret,
+    http_client: C,
+    token_retrieval_uri: Uri,
+}
+
+impl<C: HttpClient> fmt::Debug for L1TokenRetriever<C> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("L1TokenRetriever")
+            .field("client_id", &self.client_id)
+            .field("client_secret", &"<hidden>") // Do not print the secret
+            .field("token_retrieval_uri", &self.token_retrieval_uri)
+            .field("http_client", &"impl HttpClient") // HttpClient does not implement Debug
+            .finish()
+    }
+}
+
+impl<C: HttpClient> L1TokenRetriever<C> {
+    pub(super) fn new(
+        client_id: String,
+        client_secret: ClientSecret,
+        http_client: C,
+        token_retrieval_uri: Uri,
+    ) -> Self {
+        Self {
+            client_id,
+            client_secret,
+            http_client,
+            token_retrieval_uri,
+        }
+    }
+
+    fn build_request(
+        client_id: &str,
+        client_secret: &ClientSecret,
+        uri: &Uri,
+    ) -> Result<http::Request<Vec<u8>>, TokenRetrieverError> {
+        let json_body_string = json!({
```

Comment:
is this the same data as Authenticator::Request? would make sense to extend/unify it?

Reply:
Discussed over call, changed in https://github.com/newrelic/newrelic-auth-rs/pull/34/commits/d7e76e8d3f6fc916af980b6be685638c1615434a!